drivers: i2s: sai: Fix a pinmux offset calculation error

Raymond0225 opened the pull request:
Offset of pinmux base register address is how many bytes offset in device tree not words.
Fixes:#83882

Raymond0225 commented:
https://github.com/zephyrproject-rtos/zephyr/discussions/83882

decsny commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -402,7 +402,8 @@ static void enable_mclk_direction(const struct device *dev, bool dir)
 	uint32_t mask = dev_cfg->mclk_pin_mask;
-	uint32_t *base = (uint32_t *)(dev_cfg->mclk_control_base + offset);
+	/* Offset unit is byte, not word(4 bytes). */
+	uint32_t *base = (uint32_t *)((uint32_t)dev_cfg->mclk_control_base + offset);
```
this doesn't make sense, if the offset is wrong, then we should fix the offset, not cast the pointer to an integer type

Raymond0225 replied:
I would like to say, the offset is not wrong, the unit of it is byte.
but mclk_control_base is word pointer(uint32_t*).  so if you add the offset to mclk_control_base directly, the result is not what you want.
Did I make myself clear?

Raymond0225 replied:
"should be" ?   I don't think so.  I had a check on the yaml file, no more description of this parameter. It depends on the writer of device tree how to interpret it.  So far, it seems they believe that the unit is byte.
I don't think there is anything wrong.
This issue is caused by a misunderstanding of "offset" between device tree writer and i2s driver writer.

decsny commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -402,7 +402,8 @@ static void enable_mclk_direction(const struct device *dev, bool dir)
 	uint32_t mask = dev_cfg->mclk_pin_mask;
-	uint32_t *base = (uint32_t *)(dev_cfg->mclk_control_base + offset);
+	/* Offset unit is byte, not word(4 bytes). */
+	uint32_t *base = (uint32_t *)((uint32_t)dev_cfg->mclk_control_base + offset);
```
```suggestion
	uint32_t *base = dev_cfg->mclk_control_base + offset / sizeof(uint32_t);
```

decsny replied:
There is a significant difference since casting a pointer to an integer in C is bad practice due to it being implementation-specific undefined behavior, and we should not do that.

decsny reviewed (dismissed):
please add a comment to the nxp,imx-gpr.yaml describing the meanings of the pinmux specifier cells to help avoid these problems for future developers

Commit:
drivers: i2s_mcux_sai: Fix MCLK GPR address calc

The offset from devicetree is in bytes, but the i2s driver was using the
value directly in pointer arithmetic for a 32 bit pointer, leading to a
wrong calculation. Fix the expression so that the offset is converted
from a number of bytes to a number of 32 bit words as intended.

Also improve the documentation of the IOMUXC GPR binding so that it is
clear what the cells in the specifier space of "pinmux" space mean. And
change the names of the cells to be more appropriate. "offset" instead
of "pin" and "mask" instead of "function" describes more precisely what
the cells in the specifier mean.

Fixes #83882

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>
Co-authored-by: Raymond Lei <raymond.lei@nxp.com>

decsny commented:
> do you think I should update the readme file and include the description of this patch? https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/README#L127

not sure what you mean here, the readme is in the hal, this is a zephyr-only side patch, are you referring to a different patch? 

BTW, this PR currently has a merge conflict, needs a rebase

Raymond0225 commented:
I don't think we need this PR anymore, on main branch, it is fixed except the property name in yaml file and descriptions.
I close it now.
